pkg/security/admission/admission.go

```diff
@@ -0,0 +1,318 @@
+package admission
+
+import (
+	"fmt"
+	"io"
+
+	kadmission "github.com/GoogleCloudPlatform/kubernetes/pkg/admission"
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	scc "github.com/GoogleCloudPlatform/kubernetes/pkg/securitycontextconstraints"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/serviceaccount"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/watch"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/auth/user"
+	allocator "github.com/openshift/origin/pkg/security"
+	"github.com/openshift/origin/pkg/security/uid"
+)
+
+func init() {
+	kadmission.RegisterPlugin("SecurityContextConstraint", func(client client.Interface, config io.Reader) (kadmission.Interface, error) {
+		return NewConstraint(client), nil
+	})
+}
+
+type constraint struct {
+	*kadmission.Handler
+	client client.Interface
+	store  cache.Store
+}
+
+var _ kadmission.Interface = &constraint{}
+
+func NewConstraint(kclient client.Interface) kadmission.Interface {
+	store := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	reflector := cache.NewReflector(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return kclient.SecurityContextConstraints().List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(resourceVersion string) (watch.Interface, error) {
+				return kclient.SecurityContextConstraints().Watch(labels.Everything(), fields.Everything(), resourceVersion)
+			},
+		},
+		&kapi.SecurityContextConstraints{},
+		store,
+		0,
+	)
+	reflector.Run()
+
+	return &constraint{
+		Handler: kadmission.NewHandler(kadmission.Create, kadmission.Update),
+		client:  kclient,
+		store:   store,
+	}
+}
+
+// Admit determines if the pod should be admitted based on the requested security context
+// and the available SCCs.
+//
+// 1.  Find all the SCCs that the service account or user has access to
+// 2.  Full resolve each SCC
+//     1.  Determine if the user/selinux strategies requires pre-allocated values
+//     2.  Retrieve pre-allocated values from the namespace and set them on the strategies
+// 3.  Create the set of providers from the fully configured SCCs
+// 4.  Generate and validate each container's SC against the set of SCC providers.  If any container
+//     fails to validate then the entire request is rejected
+func (c *constraint) Admit(a kadmission.Attributes) error {
+	if a.GetResource() != string(kapi.ResourcePods) {
+		return nil
+	}
+
+	pod, ok := a.GetObject().(*kapi.Pod)
+	// if we can't convert then we don't handle this object so just return
+	if !ok {
+		return nil
+	}
+
+	// get all constraints that are usable by the user
```

Comment:
> 4. See if any provider allows the pod (collecting any errors while evaluating), and if so, return early. The current approach allows each container's security context to be determined by a different provider. That seems odd to me... it seems like all the containers in the pod should be validated with a single provider, all or nothing. Is there a strong argument otherwise?

Containers within a pod may not need the same level of permissions but in practice I'm starting to think that having a single SCC validate the entire pod is what will happen most of the time anyway.  The common scenario for our default SCCs will be:

1.  Admin user in a namespace has a privileged SCC and restricted SCC
1.  Admin user creates pod with no SCC requests - depending on map ordering that pod will be defaulted from either SCC UNLESS the admin makes specific requests for SC privs that error the restricted SCC.  In which case, the container validates under the admin SCC
1.  Admin user creates a pod with SCC requests - it either validates because it is in line with the restricted SCC or hits the privileged SCC which allows anything

For a pod with no requests it doesn't matter which SCC you hit, you will get defaulted and you will validate 100% of the time.  Validation failures will only occur when you start requesting specific items.

Where this breaks down is when you want to provide the ability to make non-admin multi-container pods that run with different uids, labels, etc but don't want to allow RunAsAny.  That is mitigated with a RunInRange uid strategy and we can introduce an selinux range strategy as well.  Then it is more of a management issue in terms of the number of SCCs.

Will make these updates.